📢 [프롬프트 엔지니어링, LLM(ChatGPT) 기반의 AI 서비스 개발](https://fastcampus.co.kr/data_red_golbin) 강의에서 사용하는 예제입니다.

### Playground 테스트 프롬프트

System Prompt
```
너는 사용자가 원하는 정보에 대해, 인터넷에서 정보를 수집해서 리포트를 만드는 Research Agent야.

절차는 다음과 같아.

1. 사용자가 원하는 정보에 대해 어떤 정보를 수집해야 하는지를 분석
2. 사용자가 원하는 정보를 검색
3. 검색 결과가 사용자가 원하는 정보를 제공하는데 충분한지를 확인
4. 검색 결과가 유용하지 않다면, 어떤 정보가 더 필요한지를 분석 및 선택
5. 사용자에게 제공할 정보가 충분 할 때 까지, 추가로 필요한 정보를 수집하고  확인하는 프로세스를 반복
6. 사용자에게 제공할 정보가 충분하다면, 내용을 요약해서 슬랙으로 전송

사용 할 수 있는 도구는 다음과 같아.

### Google
원하는 정보를 검색 할 때 사용.
사용법: Google("keywords")

### Report Generator
리포트를 작성할 때 사용
사용법: GenerateReport("user request")
```

User Request
```
2023년 4분기 세계 경제 전망에 대한 리포트를 작성해줘
```

Planning Prompt
```
사용자가 원하는 리포트를 만들려면 어떤 정보를 수집해야 하는지 최대 세 가지 목록을 작성해
```

Reflect Prompt
```
사용자의 요청에 따른 리포트를 작성하기에 정보가 충분한지 확인해. 정보가 충분하면 리포트 작성 도구를 이용해서 리포트를 작성해. 정보가 충분하지 않으면 검색 도구를 이용해서 정보를 검색해. 검색은 영어로 검색해.
```

Report Prompt (System)
```
너는 사용자가 원하는 정보에 대해, 인터넷에서 정보를 수집해서 리포트를 만드는 Research Agent야.

리포트의 형식은 다음과 같은 형식으로 작성해. 다음의 샘플 형식을 참고해.
"""
📝 *리포트 제목*

{리포트 내용 요약}

📑 참고자료

- <url|title>
...
"""
```

Report Prompt (User)
```
리포트 주제는 다음과 같아:
{report_title}

결과 요약은 다음과 같아:
"""
{report_summary}
"""

다음의 검색 결과를 이용해서 리포트를 작성해

--- {keyword} 검색 결과 ---
{search_result}
```

In [1]:
!pip install openai
!pip install requests
!pip install googlesearch-python

  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Created wheel for googlesearch-python: filename=googlesearch_python-1.2.3-py3-none-any.whl size=4224 sha256=4b280bab14970111e90b6c77179d62b60a607976c0afefcafbf2a08fd7f3a0bb
  Stored in directory: c:\users\ysj_1\appdata\local\pip\cache\wheels\be\78\a3\d80e85ac9551489cd9c44f59a0bc5972e79e680ce9cc6055ca
Successfully built googlesearch-python


In [2]:
from openai import OpenAI

client = OpenAI(
    api_key="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"
)

In [3]:
def research_gpt(messages, model="gpt-3.5-turbo", temperature=0, max_tokens=1024):
    system_prompt = """
너는 사용자가 원하는 정보에 대해, 인터넷에서 정보를 수집해서 리포트를 만드는 Research Agent야.

절차는 다음과 같아.

1. 사용자가 원하는 정보에 대해 어떤 정보를 수집해야 하는지를 분석
2. 사용자가 원하는 정보를 검색
3. 검색 결과가 사용자가 원하는 정보를 제공하는데 충분한지를 확인
4. 검색 결과가 유용하지 않다면, 어떤 정보가 더 필요한지를 분석 및 선택
5. 사용자에게 제공할 정보가 충분 할 때 까지, 추가로 필요한 정보를 수집하고  확인하는 프로세스를 반복
6. 사용자에게 제공할 정보가 충분하다면, 내용을 요약해서 슬랙으로 전송
"""

    functions = [
        {
            "name": "search",
            "description": "구글에서 정보를 검색합니다.",
            "parameters": {
                "type": "object",
                "properties": {
                    "keyword": {
                        "type": "string",
                        "description": "구글에 검색할 키워드",
                    }
                },
                "required": ["keyword"],
            },
        },
        {
            "name": "generate_report",
            "description": "사용자의 요청에 대한 리포트를 작성합니다.",
            "parameters": {
                "type": "object",
                "properties": {
                    "user_requests": {
                        "type": "string",
                        "description": "사용자가 요청한 리포트 주제",
                    },
                    "research_summary": {
                        "type": "string",
                        "description": "리서치 결과 요약",
                    }
                },
                "required": ["user_requests", "research_summary"],
            },
        }
    ]

    completion = client.chat.completions.create(
        model=model,
        messages=[
            {
                "role": "system",
                "content": system_prompt
            }
        ] + messages,
        functions=functions,
        function_call="auto",
        temperature=temperature,
        max_tokens=max_tokens,
    )

    return completion.choices[0].message

In [4]:
task_planning_prompt = "사용자가 원하는 리포트를 만들려면 어떤 정보를 수집해야 하는지 최대 세 가지 목록을 작성해"

research_plan = research_gpt([
    {
        "role": "user",
        "content": "2023년 4분기 세계 경제 전망에 대한 리포트를 작성해줘"
    },
    {
        "role": "user",
        "content": task_planning_prompt
    },
])

dict(research_plan)

{'content': '1. 세계 경제 전망에 대한 최신 데이터 및 통계: 2023년 4분기에 대한 세계 경제 전망을 파악하기 위해, 세계 주요 국가 및 지역의 GDP 성장률, 인플레이션률, 실업률, 수출입 현황 등의 최신 데이터와 통계를 수집해야 합니다.\n\n2. 세계 경제를 영향하는 주요 요인: 세계 경제 전망을 이해하기 위해, 세계 경제를 영향하는 주요 요인들을 파악해야 합니다. 이는 금융시장 동향, 정치적 이슈, 자연재해, 국제 무역 정책 등 다양한 요소를 포함할 수 있습니다. 이러한 요인들의 현재 상황과 예상되는 영향을 수집해야 합니다.\n\n3. 주요 산업 및 시장 동향: 세계 경제 전망을 이해하기 위해, 주요 산업 및 시장의 동향을 파악해야 합니다. 이는 금융, 에너지, 자동차, IT 등 다양한 산업과 시장을 포함할 수 있습니다. 각 산업 및 시장의 성장률, 경쟁 상황, 기술 동향 등을 수집해야 합니다.',
 'role': 'assistant',
 'function_call': None,
 'tool_calls': None}

In [6]:
print(research_plan.content)

1. 세계 경제 전망에 대한 최신 데이터 및 통계: 2023년 4분기에 대한 세계 경제 전망을 파악하기 위해, 세계 주요 국가 및 지역의 GDP 성장률, 인플레이션률, 실업률, 수출입 현황 등의 최신 데이터와 통계를 수집해야 합니다.

2. 세계 경제를 영향하는 주요 요인: 세계 경제 전망을 이해하기 위해, 세계 경제를 영향하는 주요 요인들을 파악해야 합니다. 이는 금융시장 동향, 정치적 이슈, 자연재해, 국제 무역 정책 등 다양한 요소를 포함할 수 있습니다. 이러한 요인들의 현재 상황과 예상되는 영향을 수집해야 합니다.

3. 주요 산업 및 시장 동향: 세계 경제 전망을 이해하기 위해, 주요 산업 및 시장의 동향을 파악해야 합니다. 이는 금융, 에너지, 자동차, IT 등 다양한 산업과 시장을 포함할 수 있습니다. 각 산업 및 시장의 성장률, 경쟁 상황, 기술 동향 등을 수집해야 합니다.


In [7]:
refelct_prompt = """
사용자의 요청에 따른 리포트를 작성하기에 정보가 충분한지 확인해.
정보가 충분하면 리포트 작성 도구를 이용해서 리포트를 작성해.
정보가 충분하지 않으면 검색 도구를 이용해서 정보를 검색해.
필요한 추가 검색 정보를 충분히 얻을 수 있도록 이전 검색 결과와는 다른 키워드로 검색해.
검색은 영어로 검색해.
"""

reflect_result = research_gpt([
    {
        "role": "user",
        "content": "2023년 4분기 세계 경제 전망에 대한 리포트를 작성해줘"
    },
    {
        "role": "user",
        "content": task_planning_prompt
    },
    {
        "role": "assistant",
        "content": research_plan.content
    },
    {
        "role": "user",
        "content": refelct_prompt
    },
], model="gpt-4") # 결과에 대한 사고나 평가를 위한 모델은 되도록 GPT-4를 사용

dict(reflect_result)

{'content': None,
 'role': 'assistant',
 'function_call': FunctionCall(arguments='{\n  "keyword": "2023 Q4 global economic forecast"\n}', name='search'),
 'tool_calls': None}

- function calling 실시

In [8]:
from googlesearch import search as google_search

def search(keyword: str, limit=5):
  results = google_search(keyword, num_results=limit, advanced=True)

  result_text = f"--- {keyword} 검색 결과 ---\n"
  for r in results:
      result_text += f"Title: {r.title}\n"
      result_text += f"URL: {r.url}\n"
      result_text += f"Description: {r.description}\n\n"

  return result_text

# 참고: Naver 등 다양한 검색을 쉽게 하고 싶다면 다음 유료 API를 참고하세요
# https://serpapi.com/

In [9]:
import json

arguments = json.loads(reflect_result.function_call.arguments)

search_results = search(arguments['keyword'])

print(search_results)

--- 2023 Q4 global economic forecast 검색 결과 ---
Title: Global Economic Outlook: Q4 2023
URL: https://www.euromonitor.com/article/global-economic-outlook-q4-2023
Description: Nov 17, 2023 — Global inflation is expected to decelerate to 6.9% in 2023 in the Q4 2023 forecast (7.0% in Q3). The decline in inflation is primarily due to ...

Title: Global economic outlook Q4 2023: rotating and ...
URL: https://www.woodmac.com/reports/macroeconomics-risks-and-global-trends-global-economic-outlook-q4-2023-rotating-and-rebalancing-growth-drivers-150179820/
Description: Nov 28, 2023 — We revise our GDP forecast for 2023 up to 2.6%. Crucially, we continue to hold the view that a synchronised global recession will be avoided.

Title: Global Economic Outlook
URL: https://www.conference-board.org/topics/global-economic-outlook
Description: Global real GDP is projected to grow by 3.2% in 2023, a small 0.1% upgrade from our November forecasts. Most major economies have published GDP data for the ...

Tit

In [10]:
reflect_result2 = research_gpt([
    {
        "role": "user",
        "content": "2023년 4분기 세계 경제 전망에 대한 리포트를 작성해줘"
    },
    {
        "role": "user",
        "content": task_planning_prompt
    },
    {
        "role": "user",
        "content": search_results
    },
    {
        "role": "user",
        "content": refelct_prompt
    },
], model="gpt-4") # 결과에 대한 사고나 평가를 위한 모델은 되도록 GPT-4를 사용

dict(reflect_result2)

{'content': None,
 'role': 'assistant',
 'function_call': FunctionCall(arguments='{\n  "user_requests": "2023년 4분기 세계 경제 전망에 대한 리포트",\n  "research_summary": "2023년 4분기 세계 경제 전망에 대해 다음과 같은 정보를 수집했습니다. \\n\\n1. 전 세계적으로 인플레이션은 2023년에 6.9%로 둔화될 것으로 예상되며, 이는 주로 Q3의 7.0%에서 떨어진 결과입니다. (출처: Euromonitor)\\n\\n2. 2023년에 대한 GDP 예측을 2.6%로 상향 조정했습니다. 중요한 것은 동기화된 전 세계적인 경기 침체는 피할 것이라는 견해를 계속 유지하고 있다는 점입니다. (출처: Woodmac)\\n\\n3. 전 세계 실질 GDP는 2023년에 3.2% 성장할 것으로 예상되며, 이는 11월 예측에서 0.1% 상향 조정된 것입니다. (출처: Conference Board)\\n\\n4. 미국 경제의 탄탄함에 주로 의지하여 2023년 하반기에 전 세계 경제 전망이 더욱 개선되었습니다. (출처: Euromonitor)"\n}', name='generate_report'),
 'tool_calls': None}

In [17]:
# # 이전 검색 결과에 정보가 불충분한 경우
# # function call 결과에 search 함수 사용으로 keyword 값이 출력됨
# arguments2 = json.loads(reflect_result2.function_call.arguments)

# search_results2 = search(arguments2['keyword'])

# print(search_results2)

In [ ]:
# # 검색한 결과를 추가하여 다시 생성
# result = research_gpt([
#     {
#         "role": "user",
#         "content": "2023년 4분기 세계 경제 전망에 대한 리포트를 작성해줘"
#     },
#     {
#         "role": "user",
#         "content": task_planning_prompt
#     },
#     {
#         "role": "user",
#         "content": search_results
#     },
#     {
#         "role": "user",
#         "content": search_results2
#     },
#     {
#         "role": "user",
#         "content": refelct_prompt
#     },
# ], model="gpt-4") # 결과에 대한 사고나 평가를 위한 모델은 되도록 GPT-4를 사용

# dict(result)

In [18]:
from pprint import pprint

arguments = json.loads(reflect_result2.function_call.arguments)

pprint(arguments)

{'research_summary': '2023년 4분기 세계 경제 전망에 대해 다음과 같은 정보를 수집했습니다. \n'
                     '\n'
                     '1. 전 세계적으로 인플레이션은 2023년에 6.9%로 둔화될 것으로 예상되며, 이는 주로 Q3의 '
                     '7.0%에서 떨어진 결과입니다. (출처: Euromonitor)\n'
                     '\n'
                     '2. 2023년에 대한 GDP 예측을 2.6%로 상향 조정했습니다. 중요한 것은 동기화된 전 세계적인 '
                     '경기 침체는 피할 것이라는 견해를 계속 유지하고 있다는 점입니다. (출처: Woodmac)\n'
                     '\n'
                     '3. 전 세계 실질 GDP는 2023년에 3.2% 성장할 것으로 예상되며, 이는 11월 예측에서 '
                     '0.1% 상향 조정된 것입니다. (출처: Conference Board)\n'
                     '\n'
                     '4. 미국 경제의 탄탄함에 주로 의지하여 2023년 하반기에 전 세계 경제 전망이 더욱 '
                     '개선되었습니다. (출처: Euromonitor)',
 'user_requests': '2023년 4분기 세계 경제 전망에 대한 리포트'}


In [19]:
def generate_report(title, summary, search_results, model="gpt-3.5-turbo-16k", temperature=0, max_tokens=2048):
    system_message = """
너는 사용자가 원하는 정보에 대해, 인터넷에서 정보를 수집해서 리포트를 만드는 Research Agent야.

리포트의 형식은 다음과 같은 형식으로 작성해. 다음의 샘플 형식을 참고해. 내용 요약이 길면 단락을 나눠줘.
```
📝 *리포트 제목*

{리포트 내용 요약}

📑 참고자료

- <url|title>
...
```
"""

    user_message = f"""
리포트 주제는 다음과 같아:
{title}

리서치 결과 요약은 다음과 같아:
```
{summary}
```

다음의 검색 결과를 이용해서 리포트를 작성해

{search_results}
"""

    completion = client.chat.completions.create(
        model=model,
        messages=[
            {
                "role": "system",
                "content": system_message
            },
            {
                "role": "user",
                "content": user_message
            },
        ],
        temperature=temperature,
        max_tokens=max_tokens,
    )

    return completion.choices[0].message.content

In [20]:
# report = generate_report(arguments['user_requests'],
#                 arguments['research_summary'],
#                 f"{search_results}\n\n{search_results2}")
report = generate_report(arguments['user_requests'],
                arguments['research_summary'],
                f"{search_results}")

print(report)

📝 2023년 4분기 세계 경제 전망

2023년 4분기 세계 경제 전망에 대한 리포트입니다.

2023년 4분기 세계 경제 전망에 대해 다음과 같은 정보를 수집했습니다.

1. 전 세계적으로 인플레이션은 2023년에 6.9%로 둔화될 것으로 예상되며, 이는 주로 Q3의 7.0%에서 떨어진 결과입니다. (출처: [Euromonitor](https://www.euromonitor.com/article/global-economic-outlook-q4-2023))

2. 2023년에 대한 GDP 예측을 2.6%로 상향 조정했습니다. 중요한 것은 동기화된 전 세계적인 경기 침체는 피할 것이라는 견해를 계속 유지하고 있다는 점입니다. (출처: [Woodmac](https://www.woodmac.com/reports/macroeconomics-risks-and-global-trends-global-economic-outlook-q4-2023-rotating-and-rebalancing-growth-drivers-150179820/))

3. 전 세계 실질 GDP는 2023년에 3.2% 성장할 것으로 예상되며, 이는 11월 예측에서 0.1% 상향 조정된 것입니다. (출처: [Conference Board](https://www.conference-board.org/topics/global-economic-outlook))

4. 미국 경제의 탄탄함에 주로 의지하여 2023년 하반기에 전 세계 경제 전망이 더욱 개선되었습니다. (출처: [Euromonitor](https://www.euromonitor.com/article/global-economic-outlook-q4-2023))

📑 참고자료

- [Global Economic Outlook: Q4 2023](https://www.euromonitor.com/article/global-economic-outlook-q4-2023)
- [Global economic outlook Q4 2023: rotating and 

In [21]:
import re
import requests

def send_slack(report):
    # "[제목](링크)" 형식을 슬랙의 형식인 "<링크|제목>" 으로 변환
    report = re.sub(r'\[(.*?)\]\((.*?)\)', r'<\2|\1>', report)

    slack_webhook_url = "https://hooks.slack.com/services/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"

    data = {
    'blocks': [{
        'type': 'section',
        'text': {
          'type': 'mrkdwn',
          'text': report
        }
    }]}

    requests.post(slack_webhook_url, data=json.dumps(data))

In [22]:
send_slack(report)

## 반복 플로우 자동화

1. task_planning
2. refelct
3. search
4. 2~3 반복
5. generate_report
6. slack

In [23]:
PROMPT = {
    'task_planning': """
사용자가 원하는 리포트를 만들려면 어떤 정보를 수집해야 하는지 최대 세 가지 목록을 작성해
""",
    'reflect': """
사용자의 요청에 따른 리포트를 작성하기에 정보가 충분한지 확인해.
정보가 충분하면 리포트 작성 도구를 이용해서 리포트를 작성해.
정보가 충분하지 않으면 검색 도구를 이용해서 정보를 검색해.
필요한 추가 검색 정보를 충분히 얻을 수 있도록 이전 검색 결과와는 다른 키워드로 검색해.
검색은 영어로 검색해.
"""
}

In [24]:
def task_planning(user_request):
    print(f"Planning...\n{user_request}\n\n")

    result = research_gpt([
        {
            "role": "user",
            "content": user_request
        },
        {
            "role": "user",
            "content": PROMPT['task_planning']
        },
    ])

    print(f"Research Plan:\n{result.content}\n\n")

    return result.content

In [25]:
def reflect(user_request, research_plan, search_results=[]):
    planning_messages = [
        {
            "role": "user",
            "content": user_request
        },
        {
            "role": "user",
            "content": PROMPT['task_planning']
        },
        {
            "role": "user",
            "content": research_plan
        }
    ]

    reflect_messages = [
        {
            "role": "user",
            "content": PROMPT['reflect']
        },
    ]

    context_messages = []
    for context in search_results:
      context_messages.append({
          "role": "user",
          "content": context
      })

    result = research_gpt(
        planning_messages + context_messages + reflect_messages
    , model="gpt-4") # 결과에 대한 사고나 평가를 위한 모델은 되도록 GPT-4를 사용

    return result

In [26]:
import json

def research(user_request, research_plan):
    search_memory = []

    while True:
        print("Reflect...")
        reflect_result = reflect(user_request, research_plan, search_memory)

        if reflect_result.function_call == None:
            return reflect_result.content

        arguments = json.loads(reflect_result.function_call.arguments)

        if reflect_result.function_call.name == "search":
            print("추가 정보가 필요합니다.")
            print(f"다음에 대해 검색합니다: {arguments['keyword']}\n")
            search_results = search(arguments['keyword'])
            search_memory.append(search_results)
        elif reflect_result.function_call.name == "generate_report":
            print("정보 수집이 완료되었습니다.")
            print(f"리서치 결과 요약: {arguments['research_summary']}\n")
            print("Generate...\n수집한 정보를 바탕으로 리포트를 작성합니다.\n")
            report = generate_report(
                        arguments['user_requests'],
                        arguments['research_summary'],
                        "\n\n".join(search_memory))
            break

    return report

In [27]:
user_request = "2023년 4분기 세계 경제 전망에 대한 리포트를 작성해줘"

research_plan = task_planning(user_request)

report = research(user_request, research_plan)

send_slack(report)

print(f"\nFinish:\n리포트를 작성하여 슬랙으로 전송하였습니다.")

Planning...
2023년 4분기 세계 경제 전망에 대한 리포트를 작성해줘


Research Plan:
2023년 4분기 세계 경제 전망에 대한 리포트를 작성하기 위해 수집해야 할 정보는 다음과 같습니다:

1. 세계 경제 전망: 2023년 4분기에 대한 세계 경제 전반적인 전망을 파악해야 합니다. 이를 위해 국제 기구, 금융 기관, 경제 전문가 등의 예측과 보고서를 조사해야 합니다. 세계 경제의 성장률, 인플레이션, 신용 위험, 금리 등의 주요 지표와 예측을 수집해야 합니다.

2. 주요 경제 국가의 전망: 세계 경제에 큰 영향을 미치는 주요 경제 국가들의 전망을 파악해야 합니다. 이를 위해 미국, 중국, 유럽 연합 등의 주요 경제 국가들의 경제 성장 전망, 정책 변화, 무역 전망 등을 조사해야 합니다.

3. 산업별 전망: 주요 산업 분야들의 전망을 파악해야 합니다. 이를 위해 자동차, IT, 금융, 에너지 등의 주요 산업 분야들의 성장 전망, 기술 동향, 시장 동향 등을 조사해야 합니다.

위 세 가지 정보를 수집하여 2023년 4분기 세계 경제 전망에 대한 리포트를 작성할 수 있습니다.


Reflect...
추가 정보가 필요합니다.
다음에 대해 검색합니다: 2023 Q4 global economic forecast

Reflect...
추가 정보가 필요합니다.
다음에 대해 검색합니다: 2023 Q4 US economic forecast

Reflect...
추가 정보가 필요합니다.
다음에 대해 검색합니다: 2023 Q4 China economic forecast

Reflect...
추가 정보가 필요합니다.
다음에 대해 검색합니다: 2023 Q4 industry forecast

Reflect...
정보 수집이 완료되었습니다.
리서치 결과 요약: 2023년 4분기 세계 경제 전망에 대한 리서치 결과를 요약하면 다음과 같습니다. 세계 경제는 인플레이션의 감소와 더불어 6.9%의 성장률을 보일 것으로 예상되며, 세계 GDP는 2.6% 성장할 것으로

### Function List

```python
def research_gpt(messages, model="gpt-3.5-turbo", temperature=0, max_tokens=1024):
def search(keyword, limit=5):
def generate_report(title, summary, search_results, model="gpt-3.5-turbo-16k", temperature=0, max_tokens=2048):
def send_slack(report):
def task_planning(user_request):
def reflect(user_request, research_plan, search_memory):
def research(user_request, research_plan):
```

:메모: 2023년 4분기 세계 경제 전망
2023년 4분기 세계 경제 전망에 대한 리서치 결과를 요약하면 다음과 같습니다. 세계 경제는 인플레이션의 감소와 더불어 6.9%의 성장률을 보일 것으로 예상되며, 세계 GDP는 2.6% 성장할 것으로 전망됩니다. 미국 경제는 고금리와 인플레이션 문제에 직면하고 있지만, 2023년에는 2.4%의 성장률을 보일 것으로 예상됩니다. 중국 경제는 5.0%의 성장률을 보일 것으로 예상되며, 이는 3분기 대비 0.3%포인트의 감소입니다. 산업별로 보면, 반도체 산업은 2023년 4분기에 22%의 성장률을 보일 것으로 예상되며, 이는 3분기 대비 15%포인트의 증가입니다.
:책갈피_탭: 참고자료
- Global Economic Outlook: Q4 2023
- Global economic outlook Q4 2023: rotating and rebalancing growth drivers
- Global Economic Outlook
- Global Economic Outlook Q4 2022
- World Economic Outlook, October 2023: Navigating Global
- Global Economic Forecasts: Q4 2023
- United States Economic Forecast
- Economic Forecast for the US Economy
- Fourth Quarter 2023 Survey of Professional Forecasters
- China's Economy in 2023 - A Year of Growth and Recovery
- Economic Research: Economic Outlook Asia-Pacific Q4 2023
- Global Semiconductor Manufacturing Industry Set for Q4 2023 Recovery
- Q4 2023 U.S. Markets Outlook
- Taking stock: Q4 2023 equity market outlook - Institutional
- Market Forecasts Q4 2023